### PR TITLE
Implement all floating-point primitives and instances in reference evaluator

### PR DESCRIPTION
**Pull request by RyanGlScott**
This PR is a collection of two commits which ensures that the reference evaluator's coverage of `Float` is on part with other Cryptol backends.

## Implement all floating-point primitives in reference evaluator

This ensures that all of the primitives defined in `lib/Float.cry` are supported in Cryptol's reference evaluator. Moreover, this removes the primitive table entry for `fpIsFinite`, as `fpIsFinite` is not defined as a primitive and therefore doesn't need special treatment.

I have added a `tests/regression/float_reference_eval.icry` test case which ensures that the reference evaluator does the expected thing on all floating-point operations.

Fixes https://github.com/GaloisInc/cryptol/issues/1714.

## Implement Float's Literal instance in the reference evaluator

Also extend the `float_reference_eval` test case to ensure that the reference evaluator covers all of `Float`'s instances.

Fixes https://github.com/GaloisInc/cryptol/issues/1715.

### CODE REVIEW

File: src/Cryptol/Eval/Reference.lhs [hunk 1]
```diff
@@ -1752,16 +1763,55 @@ true for all supported precisions.
 >                           y <- fromVFloat <$> yv
 >                           pure (VBit (FP.bfCompare x y == EQ))
 >
->    , "fpIsFinite" ~> vFinPoly \_ -> pure $
->                      vFinPoly \_ -> pure $
+>    , "fpIsNaN"     ~> vFinPoly \_ -> pure $
```
Comment (mccleeary-galois) [inline src/Cryptol/Eval/Reference.lhs:1766]: What set of operations are we targeting here?

> Implement all floating-point primitives

Are we wanting to support all IEEE-754-2019 operations? We are getting close to that with this set (minus the recommended ones in the standard), or is this driven by another programming language target?

**Review by mccleeary-galois — APPROVED** [review]
This is certainly an improvement of set of supported operations from previous. I did want to have a discussion about what set we should be targeting fully in the future, but isn't necessary for getting this in.

File: src/Cryptol/Eval/Reference.lhs (hunk 1 above)
Comment (RyanGlScott) [reply src/Cryptol/Eval/Reference.lhs:1766]: By "primitives", I mean all of the Cryptol primitive functions listed in [`lib/Float.cry`](https://github.com/GaloisInc/cryptol/blob/25423ce4ee65337bb75c6922e269a6d359c150e5/lib/Float.cry). (Perhaps unsurprisingly, this list of primitives closely corresponds with the operations provided by SMT-LIB's floating-point theory.)

File: src/Cryptol/Eval/Reference.lhs [hunk 2]
```diff
@@ -1707,13 +1708,23 @@ and integers.
 These functions capture the interactions with rationals.
 
 
-This just captures a common pattern for binary floating point primitives.
+These just capture common patterns for unary, binary, and trinary floating
```
Comment (yav) [inline src/Cryptol/Eval/Reference.lhs:1711]: I think the word for an operator with 3 arguments is "ternary"?

**Review by yav — APPROVED** [review]

File: src/Cryptol/Eval/Reference.lhs (hunk 2 above)
Comment (RyanGlScott) [reply src/Cryptol/Eval/Reference.lhs:1711]: You're right, I'm not sure why I though "trinary" sounded more natural. (It turns out that [trinary is a word](https://en.wiktionary.org/wiki/trinary), but much less common.) I'll change it to ternary (and rename the `*Tri*` functions to `*Tern*`).

**Pull request merged**